File: python/codegen/parsers.py
"""JSON data parsers for transliteration datasets."""

import json
import re
from typing import Any

from .models import CircledOrSquaredRecord, HyphensRecord, IvsSvsBaseRecord, RomanNumeralsRecord

__all__ = [
    "parse_simple_transliteration_records",
    "parse_hyphen_transliteration_records",
    "parse_ivs_svs_base_transliteration_records",
    "parse_kanji_old_new_transliteration_records",
    "parse_combined_transliteration_records",
    "parse_circled_or_squared_transliteration_records",
    "parse_roman_numerals_records",
]


def parse_unicode_codepoint(cp_repr: str) -> int:
    """Parse a Unicode codepoint representation like 'U+1234'."""
    match = re.match(r"^U\+([0-9A-Fa-f]+)", cp_repr)
    if not match:
        raise ValueError(f"invalid Unicode codepoint representation: {cp_repr}")
    return int(match.group(1), 16)


def codepoint_to_string(codepoint: int) -> str:
    """Convert a Unicode codepoint to a string."""
    try:
        return chr(codepoint)
    except ValueError as e:
        raise ValueError(f"invalid Unicode codepoint: {codepoint}") from e


def codepoints_to_string(codepoints: list[int]) -> str:
    """Convert a list of Unicode codepoints to a string."""
    try:
        return "".join(chr(cp) for cp in codepoints)
    except ValueError as e:
        raise ValueError(f"invalid Unicode codepoint in list: {codepoints}") from e


def parse_simple_transliteration_records(data: str) -> list[tuple[str, str]]:
    """Parse simple transliteration records from JSON data.

    :param data: JSON string containing codepoint mappings
    :returns: List of (from_char, to_char) tuples
    """
    mapping: dict[str, str | None] = json.loads(data)

    result: list[tuple[str, str]] = []
    for key, value in mapping.items():
        from_codepoint = parse_unicode_codepoint(key)
        from_char = codepoint_to_string(from_codepoint)

        if value is not None:
            to_codepoint = parse_unicode_codepoint(value)
            to_char = codepoint_to_string(to_codepoint)
        else:
            to_char = ""

        result.append((from_char, to_char))

    return result


def parse_hyphen_transliteration_records(data: str) -> list[tuple[str, HyphensRecord]]:
    """Parse hyphen transliteration records from JSON data.

    :param data: JSON string containing hyphen mapping data
    :returns: List of (code_char, HyphensRecord) tuples
    """
    records: list[dict[str, Any]] = json.loads(data)

    result: list[tuple[str, HyphensRecord]] = []
    for record in records:
        code_codepoint = parse_unicode_codepoint(record["code"])
        code_char = codepoint_to_string(code_codepoint)

        # Parse ASCII field
        ascii_val = None
        if record.get("ascii"):
            ascii_codepoints = [parse_unicode_codepoint(s) for s in record["ascii"]]
            ascii_val = codepoints_to_string(ascii_codepoints)

        # Parse JISX0201 field
        jisx0201_val = None
        if record.get("jisx0201"):
            jisx0201_codepoints = [parse_unicode_codepoint(s) for s in record["jisx0201"]]
            jisx0201_val = codepoints_to_string(jisx0201_codepoints)

        # Parse JISX0208-1978 field
        jisx0208_90_val = None
        if record.get("jisx0208-1978"):
            jisx0208_codepoints = [parse_unicode_codepoint(s) for s in record["jisx0208-1978"]]
            jisx0208_90_val = codepoints_to_string(jisx0208_codepoints)

        # Parse JISX0208-1978-windows field
        jisx0208_90_windows_val = None
        if record.get("jisx0208-1978-windows"):
            windows_codepoints = [parse_unicode_codepoint(s) for s in record["jisx0208-1978-windows"]]
            jisx0208_90_windows_val = codepoints_to_string(windows_codepoints)

        # Parse JISX0208-verbatim field
        jisx0208_verbatim_val = None
        if record.get("jisx0208-verbatim"):
            verbatim_codepoint = parse_unicode_codepoint(record["jisx0208-verbatim"])
            jisx0208_verbatim_val = codepoint_to_string(verbatim_codepoint)

        hyphen_record = HyphensRecord(
            ascii=ascii_val,
            jisx0201=jisx0201_val,
            jisx0208_90=jisx0208_90_val,
            jisx0208_90_windows=jisx0208_90_windows_val,
            jisx0208_verbatim=jisx0208_verbatim_val,
        )

        result.append((code_char, hyphen_record))

    return result


def parse_ivs_svs_base_transliteration_records(data: str) -> list[IvsSvsBaseRecord]:
    """Parse IVS/SVS base transliteration records from JSON data.

    :param data: JSON string containing IVS/SVS mapping data
    :returns: List of IvsSvsBaseRecord objects
    """
    records: list[dict[str, Any]] = json.loads(data)

    result: list[IvsSvsBaseRecord] = []
    for record in records:
        # Parse IVS field (required)
        ivs_codepoints = [parse_unicode_codepoint(s) for s in record["ivs"]]
        ivs = codepoints_to_string(ivs_codepoints)

        # Parse SVS field (optional)
        svs = None
        if record.get("svs"):
            svs_codepoints = [parse_unicode_codepoint(s) for s in record["svs"]]
            svs = codepoints_to_string(svs_codepoints)

        # Parse base90 field (optional)
        base90 = None
        if record.get("base90"):
            base90_codepoint = parse_unicode_codepoint(record["base90"])
            base90 = codepoint_to_string(base90_codepoint)

        # Parse base2004 field (optional)
        base2004 = None
        if record.get("base2004"):
            base2004_codepoint = parse_unicode_codepoint(record["base2004"])
            base2004 = codepoint_to_string(base2004_codepoint)

        ivs_svs_record = IvsSvsBaseRecord(
            ivs=ivs,
            svs=svs,
            base90=base90,
            base2004=base2004,
        )

        result.append(ivs_svs_record)

    return result


def parse_kanji_old_new_transliteration_records(data: str) -> list[tuple[str, str]]:
    """Parse kanji old-new transliteration records from JSON data.

    :param data: JSON string containing kanji mapping data
    :returns: List of (old_char, new_char) tuples
    """
    records: list[tuple[dict[str, Any], dict[str, Any]]] = json.loads(data)

    result: list[tuple[str, str]] = []
    for old_record, new_record in records:
        # Parse old kanji
        old_codepoints = [parse_unicode_codepoint(s) for s in old_record["ivs"]]
        old_char = codepoints_to_string(old_codepoints)

        # Parse new kanji
        new_codepoints = [parse_unicode_codepoint(s) for s in new_record["ivs"]]
        new_char = codepoints_to_string(new_codepoints)

        result.append((old_char, new_char))

    return result


def parse_combined_transliteration_records(data: str) -> list[tuple[str, list[str]]]:
    """Parse combined character transliteration records from JSON data.

    :param data: JSON string containing combined character mapping data
    :returns: List of (combined_char, char_list) tuples
    """
    mapping: dict[str, str] = json.loads(data)

    result: list[tuple[str, list[str]]] = []
    for key, value in mapping.items():
        from_codepoint = parse_unicode_codepoint(key)
        from_char = codepoint_to_string(from_codepoint)

        # Split the value into individual characters
        char_list = list(value)

        result.append((from_char, char_list))

    return result


def parse_circled_or_squared_transliteration_records(data: str) -> list[tuple[str, CircledOrSquaredRecord]]:
    """Parse circled or squared character transliteration records from JSON data.

    :param data: JSON string containing circled/squared character mapping data
    :returns: List of (char, CircledOrSquaredRecord) tuples
    """
    mapping: dict[str, dict[str, Any]] = json.loads(data)

    result: list[tuple[str, CircledOrSquaredRecord]] = []
    for key, record_data in mapping.items():
        from_codepoint = parse_unicode_codepoint(key)
        from_char = codepoint_to_string(from_codepoint)

        record = CircledOrSquaredRecord(
            rendering=record_data["rendering"], type=record_data["type"], emoji=record_data["emoji"]
        )

        result.append((from_char, record))

    return result


def parse_roman_numerals_records(data: str) -> list[tuple[str, RomanNumeralsRecord]]:
    """Parse roman numerals transliteration records from JSON data.

    :param data: JSON string containing roman numerals mapping data
    :returns: List of (upper_char, RomanNumeralsRecord) tuples paired by upper char
    """
    records: list[dict[str, Any]] = json.loads(data)

    result: list[tuple[str, RomanNumeralsRecord]] = []
    for record in records:
        # Parse upper and lower codes
        upper_codepoint = parse_unicode_codepoint(record["codes"]["upper"])
        upper_char = codepoint_to_string(upper_codepoint)

        lower_codepoint = parse_unicode_codepoint(record["codes"]["lower"])
        lower_char = codepoint_to_string(lower_codepoint)

        # Parse decomposed forms
        decomposed_upper = [codepoint_to_string(parse_unicode_codepoint(cp)) for cp in record["decomposed"]["upper"]]
        decomposed_lower = [codepoint_to_string(parse_unicode_codepoint(cp)) for cp in record["decomposed"]["lower"]]

        roman_record = RomanNumeralsRecord(
            value=record["value"],
            upper=upper_char,
            lower=lower_char,
            decomposed_upper=decomposed_upper,
            decomposed_lower=decomposed_lower,
        )

        result.append((upper_char, roman_record))

    return result

File: rust/src/codegen/parsers.rs
use serde::Deserialize;
use std::collections::HashMap;

use super::models::*;

#[derive(Debug, Deserialize)]
struct HyphensFileRecord {
    code: String,
    ascii: Option<Vec<String>>,
    #[serde(rename = "shift_jis")]
    _shift_jis: Option<Vec<i32>>,
    jisx0201: Option<Vec<String>>,
    #[serde(rename = "jisx0208-1978")]
    jisx0208_1978: Option<Vec<String>>,
    #[serde(rename = "jisx0208-1978-windows")]
    jisx0208_1978_windows: Option<Vec<String>>,
    #[serde(rename = "jisx0208-verbatim")]
    jisx0208_verbatim: Option<String>,
}

#[derive(Debug, Deserialize)]
struct IvsSvsSchema {
    ivs: Vec<String>,
    #[serde(rename = "svs")]
    _svs: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
struct IvsSvsFileRecord {
    ivs: Vec<String>,
    svs: Option<Vec<String>>,
    base90: Option<String>,
    base2004: Option<String>,
}

fn parse_unicode_codepoint(cp_repr: &str) -> Result<u32, anyhow::Error> {
    if let Some(captures) = cp_repr.strip_prefix("U+") {
        u32::from_str_radix(captures, 16)
            .map_err(|_| anyhow::anyhow!("invalid Unicode codepoint representation: {}", cp_repr))
    } else {
        Err(anyhow::anyhow!(
            "invalid Unicode codepoint representation: {}",
            cp_repr
        ))
    }
}

fn codepoint_to_string(codepoint: u32) -> Result<String, anyhow::Error> {
    char::from_u32(codepoint)
        .map(|c| c.to_string())
        .ok_or_else(|| anyhow::anyhow!("invalid Unicode codepoint: {}", codepoint))
}

fn codepoints_to_string(codepoints: &[u32]) -> Result<String, anyhow::Error> {
    codepoints
        .iter()
        .map(|&cp| {
            char::from_u32(cp).ok_or_else(|| anyhow::anyhow!("invalid Unicode codepoint: {}", cp))
        })
        .collect::<Result<String, _>>()
}

pub fn parse_simple_transliteration_records(
    data: &str,
) -> Result<Vec<(String, String)>, anyhow::Error> {
    let map: HashMap<String, Option<String>> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for (key, value) in map {
        let from_codepoint = parse_unicode_codepoint(&key)?;
        let from_char = codepoint_to_string(from_codepoint)?;

        let to_char = if let Some(val) = value {
            let to_codepoint = parse_unicode_codepoint(&val)?;
            codepoint_to_string(to_codepoint)?
        } else {
            String::new()
        };

        result.push((from_char, to_char));
    }

    Ok(result)
}

pub fn parse_combined_transliteration_records(
    data: &str,
) -> Result<Vec<(String, String)>, anyhow::Error> {
    let map: HashMap<String, String> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for (key, value) in map {
        let from_codepoint = parse_unicode_codepoint(&key)?;
        let from_char = codepoint_to_string(from_codepoint)?;
        // value is already a string, no need to parse it as Unicode codepoint
        result.push((from_char, value));
    }

    Ok(result)
}

pub fn parse_hyphen_transliteration_records(
    data: &str,
) -> Result<Vec<(String, HyphensRecord)>, anyhow::Error> {
    let records: Vec<HyphensFileRecord> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for record in records {
        let code_codepoint = parse_unicode_codepoint(&record.code)?;
        let code_char = codepoint_to_string(code_codepoint)?;

        let hyphen_record = HyphensRecord {
            ascii: if let Some(ascii) = record.ascii {
                let codepoints: Result<Vec<_>, _> =
                    ascii.iter().map(|s| parse_unicode_codepoint(s)).collect();
                Some(codepoints_to_string(&codepoints?)?)
            } else {
                None
            },
            jisx0201: if let Some(jisx0201) = record.jisx0201 {
                let codepoints: Result<Vec<_>, _> = jisx0201
                    .iter()
                    .map(|s| parse_unicode_codepoint(s))
                    .collect();
                Some(codepoints_to_string(&codepoints?)?)
            } else {
                None
            },
            jisx0208_90: if let Some(jisx0208_1978) = record.jisx0208_1978 {
                let codepoints: Result<Vec<_>, _> = jisx0208_1978
                    .iter()
                    .map(|s| parse_unicode_codepoint(s))
                    .collect();
                Some(codepoints_to_string(&codepoints?)?)
            } else {
                None
            },
            jisx0208_90_windows: if let Some(jisx0208_1978_windows) = record.jisx0208_1978_windows {
                let codepoints: Result<Vec<_>, _> = jisx0208_1978_windows
                    .iter()
                    .map(|s| parse_unicode_codepoint(s))
                    .collect();
                Some(codepoints_to_string(&codepoints?)?)
            } else {
                None
            },
            jisx0208_verbatim: if let Some(jisx0208_verbatim) = record.jisx0208_verbatim {
                let codepoint = parse_unicode_codepoint(&jisx0208_verbatim)?;
                Some(codepoint_to_string(codepoint)?)
            } else {
                None
            },
        };

        result.push((code_char, hyphen_record));
    }

    Ok(result)
}

pub fn parse_ivs_svs_base_transliteration_records(
    data: &str,
) -> Result<Vec<IvsSvsBaseRecord>, anyhow::Error> {
    let records: Vec<IvsSvsFileRecord> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for record in records {
        let ivs_codepoints: Result<Vec<_>, _> = record
            .ivs
            .iter()
            .map(|s| parse_unicode_codepoint(s))
            .collect();
        let ivs = codepoints_to_string(&ivs_codepoints?)?;

        let svs = if let Some(svs_vec) = record.svs {
            let svs_codepoints: Result<Vec<_>, _> =
                svs_vec.iter().map(|s| parse_unicode_codepoint(s)).collect();
            Some(codepoints_to_string(&svs_codepoints?)?)
        } else {
            None
        };

        let base90 = if let Some(base90_str) = record.base90 {
            let codepoint = parse_unicode_codepoint(&base90_str)?;
            Some(codepoint_to_string(codepoint)?)
        } else {
            None
        };

        let base2004 = if let Some(base2004_str) = record.base2004 {
            let codepoint = parse_unicode_codepoint(&base2004_str)?;
            Some(codepoint_to_string(codepoint)?)
        } else {
            None
        };

        result.push(IvsSvsBaseRecord {
            ivs,
            svs,
            base90,
            base2004,
        });
    }

    Ok(result)
}

pub fn parse_kanji_old_new_transliteration_records(
    data: &str,
) -> Result<Vec<(String, String)>, anyhow::Error> {
    let records: Vec<(IvsSvsSchema, IvsSvsSchema)> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for (old, new) in records {
        let old_codepoints: Vec<u32> = old
            .ivs
            .iter()
            .map(|s| parse_unicode_codepoint(s))
            .collect::<Result<Vec<_>, _>>()?;
        let old_str = codepoints_to_string(&old_codepoints)?;

        let new_codepoints: Vec<u32> = new
            .ivs
            .iter()
            .map(|s| parse_unicode_codepoint(s))
            .collect::<Result<Vec<_>, _>>()?;
        let new_str = codepoints_to_string(&new_codepoints)?;

        result.push((old_str, new_str));
    }

    Ok(result)
}

pub fn parse_circled_or_squared_transliteration_records(
    data: &str,
) -> Result<Vec<(String, CircledOrSquaredRecord)>, anyhow::Error> {
    let map: HashMap<String, CircledOrSquaredRecord> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for (key, record) in map {
        let from_codepoint = parse_unicode_codepoint(&key)?;
        let from_char = codepoint_to_string(from_codepoint)?;
        result.push((from_char, record));
    }

    Ok(result)
}

pub fn parse_roman_numerals_records(
    data: &str,
) -> Result<Vec<(String, RomanNumeralsRecord)>, anyhow::Error> {
    let records: Vec<RomanNumeralsRecord> = serde_json::from_str(data)?;

    let mut result = Vec::new();
    for record in records {
        // Use upper code as the key
        let upper_codepoint = parse_unicode_codepoint(&record.codes.upper)?;
        let upper_char = codepoint_to_string(upper_codepoint)?;
        result.push((upper_char, record));
    }

    Ok(result)
}
